### ISSUE TITLE
reporting spans to Jaeger backend with http transport will send an endless loop of reported spans. 

### ISSUE THREAD

**Issue by limfriend** [opentracing-contrib/java-specialagent#548]
When we configure jager's http reporting, such as "-DJAEGER_ENDPOINT = http: //myjaeger.com/api/traces", 
it will send an endless loop problem. This is because jaeger's http reporting uses the okhttp component. 
When the span is reported via okhttp, it will be captured by the okhttp plugin to generate span data, 
and when this span data is reported to the jaeger collector, new span data will be generated.

@safris  could you help to merge the pull and upgrade okhttp plugin to java-specialaget? https://github.com/opentracing-contrib/java-okhttp/pull/36

**Comment by safris** [comment]
Hi @limfriend, the SpecialAgent should not be instrumenting the tracers. There is a specific mechanism that attempts to prevent such instrumentation from happening. Issue #381 was opened regarding a similar situation with `HttpURLConnection`, and it was resolved. However, this issue, and issue #531 seem to suggest that there is still "self tracing" happening.

The right way to fix this is at a low level, which would apply to all rules, instead of just OkHttp.

In order to understand what is happening, would you be able to provide a log file with `-Dsa.log.level=FINER`?

What I'd be looking for are entries like this:

```
-------> Intercept [OkHttpAgentRule@main]: public java.util.List okhttp3.OkHttpClient.interceptors()
```

Here, `OkHttpAgentRule` is the rule class, `main` is the thread's name, and the rest is the method signature being intercepted.

If you can capture all of these entries, we'll try to understand which are coming from Jaeger, and will figure out why it's not being prevented.

**Comment by safris** [comment]
Hi @limfriend, I wanted to check in with you about this issue? Do you have a log file you can share?

**Comment by RashmiRam** [comment]
@safris I am also facing the same issue. 

This is my configuration from running a dropwizard app
`java -javaagent:/Users/user/Github/tracing/opentracing-specialagent-1.7.2.jar -Dsa.log.level=FINER  -Dsa.exporter=jaeger -DJAEGER_PROPAGATION=jaeger -DJAEGER_REPORTER_LOG_SPANS=true -DJAEGER_SAMPLER_TYPE=const -DJAEGER_SAMPLER_PARAM=1 -DJAEGER_SERVICE_NAME=myService -DJAEGER_ENDPOINT=http://localhost:14268/api/traces -jar myservice.jar server config.yml`

The following just repeats for every request to `http://localhost:14268/api/traces`

`-------> Intercept [OkHttpAgentRule@jaeger.RemoteReporter-QueueProcessor]: public java.util.List okhttp3.OkHttpClient.interceptors()
-------> Intercept [OkHttpAgentRule@jaeger.RemoteReporter-QueueProcessor]: public java.util.List okhttp3.OkHttpClient.networkInterceptors()
16:54:47.450 [jaeger.RemoteReporter-QueueProcessor] INFO io.jaegertracing.internal.reporters.LoggingReporter - Span reported: 45df5370df43822e:a54041ce75efa84c:45df5370df43822e:1 - POST
16:54:47.450 [jaeger.RemoteReporter-QueueProcessor] INFO io.jaegertracing.internal.reporters.LoggingReporter - Span reported: 45df5370df43822e:45df5370df43822e:0:1 - POST
-------> Intercept [ThriftAgentRule@jaeger.RemoteReporter-QueueProcessor]: public org.apache.thrift.protocol.TProtocol org.apache.thrift.protocol.TBinaryProtocol$Factory.getProtocol(org.apache.thrift.transport.TTransport)
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftAgentRule@jaeger.RemoteReporter-QueueProcessor]: public org.apache.thrift.protocol.TProtocol org.apache.thrift.protocol.TBinaryProtocol$Factory.getProtocol(org.apache.thrift.transport.TTransport)
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException
-------> Intercept [ThriftProtocolAgentRule@jaeger.RemoteReporter-QueueProcessor]: public void org.apache.thrift.protocol.TBinaryProtocol.writeFieldStop() throws org.apache.thrift.TException`

**Comment by safris** [comment]
Hi @RashmiRam, thank you for this log file. It seems that there is indeed a bug with the "self tracing prevention" mechanism. Would you be able to provide a sample application that demonstrates this behavior? [The code that is involved with this "self tracing prevention" mechanism is complex](https://github.com/opentracing-contrib/java-specialagent/blob/34499ee6d42c56a1a0069672f048051f84eea03b/opentracing-specialagent-api/src/main/java/io/opentracing/contrib/specialagent/AgentRule.java#L107-L131), and I would need to debug it directly in order to figure out why the `jaeger.RemoteReporter-QueueProcessor` thread is not being marked as "non instrumentable".

**Comment by safris** [comment]
@limfriend and @RashmiRam, this issue has been resolved and released with SpecialAgent v1.7.3.
